lib/developer_portal/app/controllers/developer_portal/admin/messages/trash_controller.rb

```diff
@@ -6,7 +6,7 @@ class DeveloperPortal::Admin::Messages::TrashController < DeveloperPortal::BaseC
   liquify prefix: 'messages/trash'
 
   def index
-    collection = current_account.hidden_messages.page(params[:page])
+    collection = current_account.trashed_messages.page(params[:page])
```

Comment:
This is one of the problems causing the behavior I reported.

`#hidden_messages` returns a collection of `MessageRecipient` objects, that are used for received messages (inbox), while `#trashed_messages` returns a collection of `Message` objects, that is used for sent messages (outbox).

So, returning `Message` for the received message is not correct, I'd say. Also, it seems that `hidden_at` is not set for it when the message is deleted, so that's why the `Message` drop assigns the "sent" URL to this message on the trash page:

```
if @model.hidden_at.present?
          admin_messages_trash_path(@model.id)
        else
          if @model.is_a?(MessageRecipient)
            admin_messages_inbox_path(@model.id)
          else
            admin_messages_outbox_path(@model.id)  <---- This URL is applied incorrectly
          end
        end
```

Reply:
The reason why it works if you do restore and delete again is:

1. The first time it checks for `hidden_at` on the Message model, and it's `nil`, because when deleting a **received** message, `hidden_at` is set on `MessageRecipient`, not the Message itself.

2.Then I think when "restoring" the message it actually gets deleted as a sent message, i.e. setting `hidden_at` on Message, which actually is not correct. Indeed we only need to set MessageRecipient as "hidden", not the Message itself, because we only need to remove it from this specific developer's inbox, not other recipients' inboxes.